ze/zeinterval_callbacks.cpp.erb

```diff
@@ -89,41 +92,91 @@ static void create_and_enqueue_power_message(const char* hostname, const process
 
      zeinterval_callbacks_state* state = (zeinterval_callbacks_state*) zeinterval_iter_g->callbacks_state;
      state->downstream_message_queue.push(message);
 }
 
+static void create_and_enqueue_computeEU_message(const char* hostname, const process_id_t process_id,
+                                                 const thread_id_t thread_id, const uintptr_t hDevice,
+                                                 const uint32_t subDevice, const uint64_t activeTime, const uint64_t ts) {
+    
+     zeinterval_callbacks_state* state = (zeinterval_callbacks_state*) zeinterval_iter_g->callbacks_state;
+     auto [it, inserted] = state->device_computeEngine_ref.insert({{hostname, process_id, hDevice, subDevice}, {activeTime, ts}});
+     // First entry
+     if (inserted)
+         return;
+
+     auto &[prev_activeTime, prev_ts] = it->second;
+
+     bt_message *message = create_computeEU_message(hostname, process_id,
+                                                    thread_id, hDevice, subDevice,
+                                                    ((activeTime - prev_activeTime) / static_cast<double>(ts-prev_ts)) * 1000.0,
+                                                    prev_ts,
+                                                    zeinterval_iter_g->dispatch->computeEU_event_class,
+                                                    zeinterval_self_message_iterator_g,
+                                                    zeinterval_iter_g->dispatch->stream, BACKEND_ZE);
+     state->downstream_message_queue.push(message);
+     prev_activeTime = activeTime;
+     prev_ts = ts;
+}
+
+static void create_and_enqueue_copyEU_message(const char* hostname, const process_id_t process_id, const thread_id_t thread_id,
+                                              const uintptr_t hDevice, const uint32_t subDevice, const uint64_t activeTime, const uint64_t ts) {
+    
+     zeinterval_callbacks_state* state = (zeinterval_callbacks_state*) zeinterval_iter_g->callbacks_state;
+     auto [it, inserted] = state->device_copyEngine_ref.insert({{hostname, process_id, hDevice, subDevice}, {activeTime, ts}});
+     // First entry
+     if (inserted)
+         return;
+
+     auto &[prev_activeTime, prev_ts] = it->second;
+
+     bt_message *message = create_copyEU_message(hostname, process_id,
+                                                 thread_id, hDevice, subDevice,
+                                                 ((activeTime-prev_activeTime) / static_cast<double>(ts-prev_ts)) * 1000.0,
+                                                 prev_ts,
+                                                 zeinterval_iter_g->dispatch->copyEU_event_class,
+                                                 zeinterval_self_message_iterator_g,
+                                                 zeinterval_iter_g->dispatch->stream, BACKEND_ZE);
+     state->downstream_message_queue.push(message);
+     prev_activeTime = activeTime;
+     prev_ts = ts;
+}
+
```

Comment:
Same idea as before. This code is not entirely trivial, will like to have one templated function that will be reused.

Reply:
The changes incorporated.